### PR TITLE
[serve] Handle `None` in `ReplicaConfig`'s `resource_dict`

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

At least two lines of code rely on none of `ReplicaConfig`'s [`resource_dict`](https://github.com/ray-project/ray/blob/bdd3b9a0abdb96f70584092ae4d97f74769533b3/python/ray/serve/config.py#L231) values being `None` ([line 1](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L759), [line 2](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L269)). However, #23619 made `"memory"` in the `resource_dict` [default to `None`](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/config.py#L279).

This causes flakiness whenever either of the two lines are executed. In particular, line 1 is called in the [`resource_requirements()`](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L748) function, which is only called in [`DeploymentState`](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L906)'s [`_check_and_update_replicas()`](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L1388) function whenever [a replica is slow to start](https://github.com/ray-project/ray/blob/40646eecd46f3072dac425320ab966548f59fd96/python/ray/serve/deployment_state.py#L1453). This makes updates flaky when a replica starts slowly ([example](https://console.anyscale.com/o/anyscale-internal/projects/prj_2xR6uT6t7jJuu1aCwWMsle/clusters/ses_52YsYqbP5FrLiJk2SKyUEtqg?command-history-section=command_history)).

Line 2 runs for all deployments; however, it uses a generator function, so it only causes an error if both `"num_cpus"` and `"num_gpus"` are non-zero but `"memory"` is `None`. Since `"num_gpus"` has a default of 0, this is somewhat unlikely.

This change sets `"memory"`'s default to `0` in the `resource_dict` but keeps the default as `None` in `ray_actor_options`. It adds logic to both problematic lines to handle `None` in case of future settings updates. It also adds unit tests to prevent regressions.

## Related issue number

Addresses flakiness in #23747.

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - New unit tests are added to prevent regression.

### CODE REVIEW

File: python/ray/serve/deployment_state.py [hunk 1]
```diff
@@ -264,9 +264,11 @@ def start(self, deployment_info: DeploymentInfo, version: DeploymentVersion):
         )
 
         self._actor_resources = deployment_info.replica_config.resource_dict
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/deployment_state.py:266]: The context for that is in #23619. Essentially, the `None` is passed to Ray, so deployments use the Ray actor memory default.

Internally, Serve treats the default memory as `0` (even though it's actually using Ray's default) to indicate that the deployment doesn't have any memory requirements. In reality, Ray doesn't allow actors to set memory to `0`.

That's why in this change we set `memory` to `0` in the `resource_dict`, but we use `None` in `ray_actor_options`.

File: python/ray/serve/tests/test_deployment_state.py [hunk 2]
```diff
@@ -2086,5 +2086,23 @@ def compare(before, after):
     compare([2, 2, 3, 3], [2, 2, 3, 3])  # if equal, ordering should be kept
 
 
+def test_resource_requirements_none(mock_deployment_state):
+    """Ensure resource_requirements doesn't break if a requirement is None"""
+
+    from ray.serve.deployment_state import DeploymentReplica
```
Comment (edoakes) [inline python/ray/serve/tests/test_deployment_state.py:2092]: import at top of file

**Review by edoakes — APPROVED** [review]
Thanks for the fix.

It looks like you're currently both setting the default to `0` and handling the `None` case in the offending code. Do we need to be doing both here?

File: python/ray/serve/tests/test_deployment_state.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/test_deployment_state.py]: Thanks, moved the import.

**Comment by suquark** [comment]
My question: why not just skips the `None` options? I mean we should not even keep them in the directory. After https://github.com/ray-project/ray/pull/23127, all options not specified in `@ray.remote` or `.options()` will stay unspecified.

**Comment by shrekris-anyscale** [comment]
> why not just skips the None options?

@suquark What do you mean by skip the `None` options? Do you mean that we should stop validating them at the Serve level, or that we should leave them out of the `resource_dict`? As far as I can tell, this code only validates the `ray_actor_options` and provides default values in case the user explicitly specifies any option as `None`.

**Comment by suquark** [comment]
> > why not just skips the None options?
> 
> @suquark What do you mean by skip the `None` options? Do you mean that we should stop validating them at the Serve level, or that we should leave them out of the `resource_dict`? As far as I can tell, this code only validates the `ray_actor_options` and provides default values in case the user explicitly specifies any option as `None`.

Yes, but finally the options are passed to Ray actors and tasks, right? It is necessary to set `num_cpus` to 1, but you can just leave out other arguments. If they does not exist in `ray_actor_options`, do not `self.ray_actor_options.setdefault("memory", None)`. Also we should check the arguments here, but we can use https://github.com/ray-project/ray/blob/f400c20246218d933b54d4f21f2117581116cdc5/python/ray/_private/ray_option_utils.py#L193 instead. The current check is incomplete.

**Review by edoakes — APPROVED** [review]
LGTM as a localized bugfix for the issue we observed in release tests.

@shrekris-anyscale I think we should follow up with @suquark's suggestion to improve this as we update the validation & schema logic for deployment graphs.

**Pull request merged**